### ISSUE TITLE
Proposal for Enhancing Session Security in Fat-Free Framework

### ISSUE THREAD

**Issue by Division70** [f3-factory/fatfree-core#383]
Dear F3 Developer Community,

I'd like to propose an enhancement to the `read()` function in the session handling mechanism of Fat-Free Framework. The current implementation, while effective, may be too rigid for certain use cases, particularly in mobile environments where IP addresses can change frequently.

Here's a summary of the proposed changes:

1. Implement a threat level assessment system instead of a binary check.
2. Allow for customizable handling of high-threat situations.
3. Improve logging for suspicious sessions.

Key features of the proposed implementation:

- A point-based threat level system that considers IP changes, user agent changes, and session inactivity.
- Configurable thresholds for threat levels and inactivity timeouts.
- A hook for custom high-threat handling, falling back to the default 403 error if not defined.
- Enhanced logging for low to medium threat levels.

This approach aims to balance security with flexibility, allowing developers to fine-tune the session security according to their specific needs while maintaining a robust default behavior.

I welcome any feedback, suggestions, or concerns from the community. Thank you for considering this proposal.

Best regards,

Christophe ROLLAND

Here's a snippet of the core changes:

	const THREAT_LEVEL_HIGH = 3;
	const INACTIVITY_THRESHOLD = 3600;

	/**
	*	Return session data in serialized format
	*	@return string
	*	@param $id string
	**/
	function read($id) {
		$this->sid=$id;
		if (!$data=$this->_cache->get($id.'.@'))
			return '';
		$this->_data = $data;
		if ($data['ip']!=$this->_ip || $data['agent']!=$this->_agent) {
			$fw=Base::instance();
			$threatLevel = 0;

			// Vérify IP
		    if ($this->_ip != $data['ip']) {
		        $threatLevel += 1;
		    }

		    // Vérify Agent
		    if ($this->_agent != $data['agent']) {
		        $threatLevel += 2;
		    }

		    // Vérify timestamp
		    $lastActivity = $this->stamp();
		    $currentTime = time();
		    $timeDiff = $currentTime - $lastActivity;
		    if ($timeDiff > self::INACTIVITY_THRESHOLD) { // more than 1 hour
		        $threatLevel += 1;
		    }

		    // Simple threat level evaluator
		    if ($threatLevel >= self::THREAT_LEVEL_HIGH) {
		        $this->destroy($id);
				$this->close();
				unset($fw->{'COOKIE.'.session_name()});

				// if a custom user function exist
			    if (function_exists('handleHighThreatLevel')) {
			        // call custom user function
			        handleHighThreatLevel($threatLevel, $this->_data);
			    } else {
			        // no custom fonctions ? let's go to the 403
			        $fw->error(403);
			    }

		        return false;
		    } elseif ($threatLevel > 0) {
        		// log lower threat level
        		$logger = new \Log('logs/session_suspects.log');
        		$logger->write("Session suspect : IP={$this->_ip}, Agent={$this->_agent}, ThreatLevel=$threatLevel, Time=$currentTime");

        		// Update session data
		        $data['ip'] = $this->_ip;
		        $data['agent'] = $this->_agent;
		        $this->_cache->set($id.'.@', $data);

			}
		}
		return $data['data'];
	}

	// anymwhere on the app
	function handleHighThreatLevel($threatLevel, $data) {
  		// custom user fonction to manage the behavior

    	// custom redirect or anything you wan
    	header('Location: /error/high-threat');
    	exit;
	}

**Comment by Division70** [comment]
**Additional feedback after real-world usage**

 After several weeks of production usage, I wanted to add some concrete feedback on this proposal.

The enhanced session handling logic has now been running in real conditions (front-office traffic, mobile users, crawlers, social media previews) with **more than 4,000 suspicious session events logged and analyzed**.

 Key observations:

> * **0 high-threat events (TL3)** were detected
> * ~70% low-threat events (TL1)
> * ~30% medium-threat events (TL2)
> * The vast majority of these events come from:
>
>   * Mobile networks with frequent IP rotation
>   * Multi-IP crawlers (Googlebot in particular)
>   * Social media in-app browsers (notably Facebook / MetaView), which frequently alter headers and IPs during the same navigation flow

Before applying this enhancement, the framework produced a significant amount of false-positive **403 responses**, especially visible in Google Search Console crawl reports.
After deploying this threat-level approach, **crawl errors dropped from ~9% to 0%**, without weakening session security.

This confirms that the current binary IP/agent validation, while secure, is **too strict for front-office usage in modern environments** (mobile, CG-NAT, multi-IP crawling).

 The proposed threat-level system allows:

* **Strict policies for back-office contexts** (e.g. threshold = 1, no tolerance)
* **Adaptive policies for front-office contexts**, preventing legitimate users and crawlers from being ejected due to infrastructure behavior beyond their control

In practice, this enhancement avoids the common workaround of *commenting out session checks*, and instead provides a controlled, observable, and configurable security model.

Based on these results, I strongly believe this approach should be **preferred for front-office session handling**, while still allowing strict configurations where required.

Thank you again for considering this improvement.

**Comment by ikkez** [comment]
Hi. Thanks for the feedback. I was just working on session handler for v4 and refactored it now to have a scoring based threat detection, similar as you proposed. It goes like this now:

```php
$session = new \F3\Session(); // cache-based session handler
$session->threatLevelThreshold = 2;
$session->inactivityThreshold = 3600 * 2;
$session->onRead = function (\F3\Session $session, int $threatLevel) {
    // logging
    if ($threatLevel > 0) {
        // log lower threat level
        $logger = new \Log('logs/session_suspects.log');
        $logger->write("Session suspect : IP={$session->ip()}, Agent={$session->agent()}, ThreatLevel=$threatLevel");
    }
};
$session->onSuspect = function (\F3\Session $session, string $sessionId) use($f3) {
    $this->destroy($id);
    $this->close();
    unset($f3->{'COOKIE.'.session_name()});
    // do something more
};
// ready
$f3->set('SESSION.foo', 'bar'); // <-- start session automatically
```

This way it should be possible to customise the behaviour more easily. What do you think?

**Comment by Division70** [comment]
Hi ikkez,

Thanks a lot for this refactoring, this is exactly the direction I was hoping for.

The scoring-based approach combined with configurable thresholds fully addresses the false positives observed on mobile networks, multi-IP crawlers and in-app browsers.

 I’ll be happy to test this version in real conditions and share feedback if useful.

 I noticed that you kept a fixed weighting in `getThreatLevel()`, which is a very reasonable default choice for the core.
For applications that require more granularity, one possible approach could be to override the threat scoring logic at application level, for example:

> ```php
> class MySession extends \F3\Session
> {
>     protected function getThreatLevel(string $knownIP, string $knownAgent): int
>     {
>         $tl = 0;
>
>         // expired session: informational only
>         if (time() - $this->stamp() > $this->inactivityThreshold)
>             $tl += 0;
>
>         // IP change
>         if ($knownIP !== $this->ip())
>             $tl += 2;
>
>         // User-Agent change
>         if ($knownAgent !== $this->agent())
>             $tl += 3;
>
>         return $tl;
>     }
> }
> ```

This keeps the core simple while allowing application-specific tuning where needed.
Does this approach make sense to you ?

**Comment by ikkez** [comment]
Yeah absolutely.. 👍  you can always extend and overwrite specific parts,.. I had that in mind while building it.
